api/envoy/extensions/filters/http/rate_limit_quota/v3/rate_limit_quota.proto

```diff
@@ -0,0 +1,159 @@
+  // immediately return success.
+  config.core.v3.ConfigSource rlqs_server = 1;
+
+  // The application domain to use when calling the service.
+  string domain = 2;
+
+  // Unified matchers.
+  xds.type.matcher.v3.Matcher bucket_matchers = 3;
+}
+
+// [#next-free-field: 7]
+message RateLimitQuotaBucketSettings {
+  // Configures the behavior when the first request has been matched to
+  // the bucket, and the data plane is waiting for the RLQS to return the
+  // first quota assignment.
+  message NoAssignmentBehavior {
+    oneof behavior {
+      option (validate.required) = true;
+
+      // Allow or Deny the requests until the bucket is assigned a quota.
+      type.v3.RateLimitStrategy.BooleanChoice process_requests = 1;
+
+      // Queue the requests until the bucket is assigned a quota.
+      // The data plane maintains a separate queue for each bucket that has
+      // pending requests.
+      // When the server returns quota assignment, the data plane begins
+      // dequeuing the requests in the FIFO order, and processing them according
+      // to the assignment. No individual request/response matching needed.
+      //
+      // .. note::
+      // To avoid memory exhaust, queue configuration allows to
+      // to impose constraints on the maximum number of pending requests
+      // in the queue, and choose what to do with the request over the limit.
+      type.v3.RateLimitStrategyQueue queue_requests = 2;
+    }
+  }
+
+  // Specifies the behavior when the bucket's assignment has expired, and
+  // hasn't been refreshed due to server's unavailability.
+  message FallbackBehavior {
+    message FallbackLastAssignment {
+      // Limit the time the last assignment is used as a fallback.
+      // [#comment:TODO(sergiitk): what to do after that?]
+      google.protobuf.Duration time_to_live = 1;
+    }
+
+    oneof behavior {
+      option (validate.required) = true;
+
+      // Pre-configured local rate limit strategy.
+      type.v3.RateLimitStrategy fallback_rate_limit = 1;
+
+      // Reuse last quota assignment associated to the bucket.
+      FallbackLastAssignment fallback_last_assignment = 2;
+    }
+  }
+
+  message DenyResponseSettings {
+    // Status code to use to fail requests when quota is exceeded.
+    // [#comment:TODO: figure out the right defaults for these fields!]
+    int32 http_status_code = 1;
+    // For HTTP requests.
+
+    google.rpc.Status deny_grpc_status = 2;
+    // For gRPC requests.
+
+    // A response body to send to the downstream client when the response code
+    // is not OK.
+    bytes http_raw_body = 3;
+
+    // A list of headers to add to the response when the request is throttled.
+    repeated config.core.v3.HeaderValue response_headers_to_add = 4;
+  }
+
+  message BucketIdBuilder {
+    message ValueBuilder {
+      oneof value_specifier {
+        string string_value = 1;
+
+        config.core.v3.TypedExtensionConfig custom_value_builder = 2;
```

Comment:
Please document what kinds of plugins you would expect to see here.

Reply:
I think the normal way of configuring an extension is to use `TypedExtensionConfig`, even in cases where the `name` field isn't actually needed.

I don't think there can be multiple custom_value_builder fields for a given key.  The value of a given key can come from only one place.

Reply:
Yep. Make sense. `TypedExtensionConfig` is preferred way. Thanks!

Reply:
@markdroth - after carefully considering our options here, I don't know if there's a good extension category fit right now.

1) The next best thing is `envoy.matching.inputs` - but the problem with them is they are inputs for matching - and may not always produce a string output.
2) There's `envoy.matching.common_inputs` - it's close to `envoy.matching.inputs`, but not only contains `envoy.matching.common_inputs.environment_variable`. This is also a matcher, and may not be a good choice security-wise.
3) `RegexMatchAndSubstitute` is close too, but it doesn't have the category.

We potentially may need a separate category. cc @kyessenov

Reply:
I think the HTTP matching input category would work here. Per https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/advanced/matching/matching_api#http-input-functions, common inputs and some network inputs apply in the HTTP context in addition to the expected header value inputs.

Reply:
Thank you @kyessenov - implemented in https://github.com/envoyproxy/envoy/pull/19793/commits/8c20ec1384444fefd433a3b839ee0bc2f78bfed5. Rendered doc looks great!

<img width="961" alt="image" src="https://user-images.githubusercontent.com/672669/166974372-8609d8f3-bac2-46fe-928b-cd0bfb526b58.png">